src/scheduler/MpiWorld.cpp

```diff
@@ -374,28 +433,32 @@ int MpiWorld::irecv(int sendRank,
                     faabric::MPIMessage::MPIMessageType messageType)
 {
     int requestId = (int)faabric::util::generateGid();
+    auto it = reqIdToRanks.try_emplace(requestId, sendRank, recvRank);
+    if (!it.second) {
+        SPDLOG_ERROR("Request ID {} is already present in map", requestId);
+        throw std::runtime_error("Request ID already in map");
+    }
+
+    faabric::scheduler::UnackedMessageBuffer::Arguments args = {
+        sendRank, recvRank, buffer, dataType, count, messageType
+    };
 
-    std::promise<void> resultPromise;
-    std::future<void> resultFuture = resultPromise.get_future();
-    threadPool->getMpiReqQueue()->enqueue(
-      std::make_tuple(requestId,
-                      std::bind(&MpiWorld::recv,
-                                this,
-                                sendRank,
-                                recvRank,
-                                buffer,
-                                dataType,
-                                count,
-                                nullptr,
-                                messageType),
-                      std::move(resultPromise)));
-
-    // Place the promise in a map to wait for it later
-    futureMap.emplace(std::make_pair(requestId, std::move(resultFuture)));
+    getUnackedMessageBuffer(sendRank, recvRank)->ids.push_back(requestId);
+    getUnackedMessageBuffer(sendRank, recvRank)->args.push_back(args);
+    getUnackedMessageBuffer(sendRank, recvRank)->msgs.push_back(nullptr);
```

Comment:
Given that we've got a separate object to represent an `UnackedMessageBuffer` it seems like we could wrap these three lines into a method on that object, e.g. `addMessage(requestId, args, nullptr)`.